File: data-structures/lists/java/nested_lists_find_spacecraft.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

import java.util.ArrayList;
import java.util.Scanner;

class NestedLists  {
    // Index values of each astronaut
    public static final int CRAFT_INDEX = 0;
    public static final int NAME_INDEX = 1;
    public static final int DAYS_INDEX = 2;

    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Declare and initialise the nested list of astronauts
        ArrayList<ArrayList<String>> allAstronauts = new ArrayList<ArrayList<String>>();
        initialiseAstronauts(allAstronauts);

        // Use Scanner for reading user input from the console
        Scanner scanner = new Scanner(System.in);

        // Find all the astronauts on a spacecraft
        System.out.println("Get a list of all astronauts that are on a specific spacecraft");
        System.out.println("Enter the spacecraft name:");
        String spacecraft = scanner.nextLine();
        ArrayList<ArrayList<String>> foundAstronauts = getAstronautsBySpacecraft(allAstronauts, spacecraft);

        // Output the search results
        if (foundAstronauts.size() == 0) {
            System.out.println("\nNo astronauts found in the spacecraft " + spacecraft);
        }
        else {
            System.out.println("\nAstronauts found:");
            displayAllAstronauts(foundAstronauts);
        }
    }

    // Return a list of astronauts that are on a given spacecraft
    public static ArrayList<ArrayList<String>> getAstronautsBySpacecraft(
            ArrayList<ArrayList<String>> allAstronauts, String givenSpacecraft) {

        // Declare a list to store the found astronauts in
        ArrayList<ArrayList<String>> foundList = new ArrayList<ArrayList<String>>();

        // Repeat for all astronauts in the outer list using the index i
        for (int i = 0; i < allAstronauts.size(); i++) {
            // Check if the astronaut is in the spacecraft
            if (givenSpacecraft.equals(allAstronauts.get(i).get(CRAFT_INDEX))) {
                foundList.add(allAstronauts.get(i));
            }
        }

        return foundList;
    }

    // Display the values of each astronaut in the given list
    public static void displayAllAstronauts(ArrayList<ArrayList<String>> allAstronauts) {
        // Repeat for each astronaut in the outer list
        for (ArrayList<String> astronaut : allAstronauts) {
            // Repeat for each value in the inner list
            for (String value : astronaut) {
                System.out.print(value + "  "); // Do not print new line
            }
            System.out.println(); // Print a new line after each astronaut
        }
    }

    // Initialise the nested list of astronauts
    public static void initialiseAstronauts(ArrayList<ArrayList<String>> allAstronauts) {
        ArrayList<String> first = new ArrayList<String>();
        first.add("ISS");
        first.add("Sergey Prokopyev");
        first.add("176");
        allAstronauts.add(first);

        ArrayList<String> second = new ArrayList<String>();
        second.add("ISS");
        second.add("Dmitry Petelin");
        second.add("176");
        allAstronauts.add(second);

        ArrayList<String> third = new ArrayList<String>();
        third.add("ISS");
        third.add("Frank Rubio");
        third.add("176");
        allAstronauts.add(third);

        ArrayList<String> fourth = new ArrayList<String>();
        fourth.add("Tiangong");
        fourth.add("Fei Junlong");
        fourth.add("112");
        allAstronauts.add(fourth);

        ArrayList<String> fifth = new ArrayList<String>();
        fifth.add("Tiangong");
        fifth.add("Deng Qingming");
        fifth.add("107");
        allAstronauts.add(fifth);

        ArrayList<String> sixth = new ArrayList<String>();
        sixth.add("Tiangong");
        sixth.add("Zhang Lu");
        sixth.add("107");
        allAstronauts.add(sixth);

        ArrayList<String> seventh = new ArrayList<String>();
        seventh.add("ISS");
        seventh.add("Stephen Bowen");
        seventh.add("54");
        allAstronauts.add(seventh);

        ArrayList<String> eighth = new ArrayList<String>();
        eighth.add("ISS");
        eighth.add("Warren Hoburg");
        eighth.add("14");
        allAstronauts.add(eighth);

        ArrayList<String> ninth = new ArrayList<String>();
        ninth.add("ISS");
        ninth.add("Sultan Al Neyadi");
        ninth.add("14");
        allAstronauts.add(ninth);

        ArrayList<String> tenth = new ArrayList<String>();
        tenth.add("ISS");
        tenth.add("Andrey Fedyaev");
        tenth.add("14");
        allAstronauts.add(tenth);
    }
}

File: data-structures/lists/visual-basic/nested_lists_find_spacecraft.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program

Module Program
    ' Index values of each astronaut
    Public Const CraftIndex As Integer = 0
    Public Const NameIndex As Integer = 1
    Public Const DaysIndex As Integer = 2

    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        ' Declare and initialise the astronaut list
        Dim allAstronauts As New List(Of List(Of String))({
            New List(Of String)({"ISS", "Sergey Prokopyev", "176"}),
            New List(Of String)({"ISS", "Dmitry Petelin", "176"}),
            New List(Of String)({"ISS", "Frank Rubio", "176"}),
            New List(Of String)({"Tiangong", "Fei Junlong", "112"}),
            New List(Of String)({"Tiangong", "Deng Qingming", "107"}),
            New List(Of String)({"Tiangong", "Zhang Lu", "107"}),
            New List(Of String)({"ISS", "Stephen Bowen", "54"}),
            New List(Of String)({"ISS", "Warren Hoburg", "14"}),
            New List(Of String)({"ISS", "Sultan Al Neyadi", "14"}),
            New List(Of String)({"ISS", "Andrey Fedyaev", "14"})
        })

        ' Find all the astronauts on a spacecraft
        Console.WriteLine("Get a list of all astronauts that are on a specific spacecraft")
        Console.WriteLine("Enter the spacecraft name:")
        Dim spacecraft As String = Console.ReadLine()
        Dim foundAstronauts As List(Of List(Of String)) = GetAstronautsBySpacecraft(allAstronauts, spacecraft)

        ' Output the search results
        If foundAstronauts.Count = 0 Then
            Console.WriteLine(vbLf & $"No astronauts found in the spacecraft {spacecraft}")
        Else
            Console.WriteLine(vbLf & "Astronauts found:")
            DisplayAllAstronauts(foundAstronauts)
        End If
    End Sub

    ' Return a list of astronauts that are on a given spacecraft
    Function GetAstronautsBySpacecraft(ByVal allAstronauts As List(Of List(Of String)), ByVal givenSpacecraft As String) As List(Of List(Of String))
        ' Declare a list to store the found astronauts in
        Dim foundList As New List(Of List(Of String))()

        ' Repeat for all astronauts in the outer list using the index i
        For i = 0 To allAstronauts.Count - 1
            ' Check if the astronaut is in the spacecraft
            If givenSpacecraft.Equals(allAstronauts(i)(CraftIndex)) Then
                foundList.Add(allAstronauts(i))
            End If
        Next

        Return foundList
    End Function


    ' Display the values of each astronaut in the given list
    Sub DisplayAllAstronauts(ByVal allAstronauts As List(Of List(Of String)))
        ' Repeat for each astronaut in the outer list
        For Each astronaut In allAstronauts
            ' Repeat for each value in the inner list
            For Each value In astronaut
                Console.Write($"{value}  ") ' Do not print new line
            Next
            Console.WriteLine() ' Print a new line after each astronaut
        Next
    End Sub
End Module